python/lsst/sims/catUtils/mixins/PhotometryMixin.py

```diff
@@ -148,6 +148,44 @@ def get_lsst_photometric_uncertainties(self):
                                                  'lsstBandpassDict')
 
 
+    def calculateVisibility(self, magFilter, sigma=0.1, randomSeed=None, pre_generate_randoms=False):
+        """
+        Determine (probabilistically) whether a source was detected or not.
+
+        The 'completeness' of source detection at any magnitude is calculated by
+          completeness = (1 + e^^(magFilter-m5)/sigma)^(-1)
+        For each source with a magnitude magFilter, if a uniform random number [0-1)
+          is less than or equal to 'completeness', then it is counted as "detected".
+        See equation 24, figure 8 and table 5 from SDSS completeness analysis in
+        http://iopscience.iop.org/0004-637X/794/2/120/pdf/apj_794_2_120.pdf
+        "THE SLOAN DIGITAL SKY SURVEY COADD: 275 deg2 OF DEEP SLOAN DIGITAL SKY SURVEY IMAGING ON STRIPE 82"
+
+        @ param [in] magFilter is the magnitude of the object in the observed filter.
+
+        @ param [in] sigma is the FWHM of the distribution (default = 0.1)
+
+        @ param [in] randomSeed is an option to set a random seed (default None)
+        @ param [in] pre_generate_randoms is an option (default False) to pre-generate a series of 12,000,000 random numbers
+           for use throughout the visibility calculation [the random numbers used are randoms[objId]].
+
+        @ param [out] visibility (None/1).
+        """
+        if len(magFilter) == 0:
+            return numpy.array([])
+        # Calculate the completeness at the magnitude of each object.
+        completeness = 1.0 / (1 + numpy.exp((magFilter - self.obs_metadata.m5[self.obs_metadata.bandpass])/sigma))
+        if randomSeed is not None:
+            numpy.random.seed(randomSeed)
+        if pre_generate_randoms:
```

Comment:
if you make a database query that is large enough to come back in multiple chunks, the randoms will be generated each time.

This should probably be

if pre_generate_randoms and not hasattr(self, 'randoms')

Also: maybe rename 'randoms' to 'ssm_randoms', just in case we end up adopting this solution for another use case